bundles/org.eclipse.equinox.p2.director/src/org/eclipse/equinox/internal/p2/director/Projector.java

```diff
@@ -1083,4 +1083,11 @@ private void rememberHostMatches(IInstallableUnitFragment fragment, List<IInstal
 	public void setUserDefined(boolean containsKey) {
 		userDefinedFunction = containsKey;
 	}
+
+	public void close() {
```

Comment:
Maybe projector should implement Autocloseable then?

Reply:
If you look at where close is called in a several places for the one place that a Projector is created, it wouldn't really help make anything simpler. at least in the p2 code itself.

Reply:
But why is close called in seveveral places and not once after the user is done? AutoColeable has the advantage that eclipse warns you if you do not close it after use.

Reply:
The complete code is there for your inspection to answer questions like this yourself...  

The Projector is used in a rather ugly way.   In some cases the Projector is used to create and return a plan, at which point it's no longer used, and can be closed.  In other cases it's returned from the method (whose return type is Object) and must be closed by the caller.  I haven't properly covered all the uses of the call to getSolutionFor either, just the path followed by the marketplace tester.  And the marketplace tester job still dies with the newer version of sat4j which requires further investigation, but testing a job that runs for hours that maybe has a leak is not trivial...

Reply:
>  In other cases it's returned from the method (whose return type is Object) and must be closed by the caller. 

Maybe it then cloud return an Closable instead to indicate the to the caller then or convert it? This are just some ideas without having to investigate this in details, but having a "close" method that is arbitrary called really seems like code-smell ... especially if that not really "closes" but seem to reset some internal state...

If I search for Projector#dependencyHelper it seems only be used inside the projector and there is only one public method referencing this, so I think one should remove the field, and pass the object to subcalls as a parameter instead and always reset it (or just stop the job if possible) before returning would prevent the instance from being hard referenced here.

Reply:
e.g. `DependencyHelper#stopSolver()` seems a possible candidate we should always call in a finally block.

Reply:
Yes, the code smells.   But I really have so many other things to keep me busy that it's hard to keep track of them all.  And that's not to mention the open bugs in p2 that, in my opinion, are generally all more important than removing this particular smell, at the risk of adding this this list:

https://bugs.eclipse.org/bugs/buglist.cgi?bug_status=UNCONFIRMED&bug_status=NEW&bug_status=ASSIGNED&bug_status=REOPENED&classification=Eclipse%20Project&component=p2&list_id=21260880&order=bug_id%20DESC&product=Equinox&query_format=advanced

Perhaps you could focus on one of those because then some real live human being will see some real actual improvement in their lives.

Reply:
If my goal would be to make people happy I won't review code ;-)

It seems that this could really cause some serious problem making "real live human being" unhappy, so I think its worth to have a sustained fix for this, of course you could simply ignore that and wait until you or some else run into the same problem because there is somewhere another `close()` missing ...